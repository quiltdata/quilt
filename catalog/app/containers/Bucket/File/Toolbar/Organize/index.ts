export * as Context from './Context'
export { default as Options } from './Options'
