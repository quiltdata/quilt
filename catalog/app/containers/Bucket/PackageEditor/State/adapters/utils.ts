import type { FileWithPath } from 'react-dropzone'

import { Status } from 'components/FileManager/FileRow'
import type { TreeEntry } from 'components/FileManager/FileTree'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export function calcChildren(
  entry: Model.PackageEntry | Model.S3File | FileWithPath,
  tailParts: string[],
  children: TreeEntry[] = [],
): TreeEntry[] {
  const [name, ...tail] = tailParts
  const found = children.find((child) => child.id === name)
  if (found) {
    if (tail.length) {
      found.children = calcChildren(entry, tail, found.children)
    }
    return children
  }

  return children.concat({
    id: name,
    name: tail.length ? s3paths.ensureSlash(name) : name,
    size: tail.length ? 0 : entry.size,
    status: Status.Unchanged,
    children: tail.length ? calcChildren(entry, tail, []) : undefined,
  })
}

export function sortEntries(entries: TreeEntry[]): TreeEntry[] {
  return [...entries].sort((a, b) => {
    if (a.children && !b.children) return -1
    if (!a.children && b.children) return 1
    return a.name.localeCompare(b.name)
  })
}
