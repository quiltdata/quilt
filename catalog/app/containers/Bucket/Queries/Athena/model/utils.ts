export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no data. It is not initialized
// `Loading` is loading
// `Error` is error
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

export function wrapData<T>(data: Data<T>, setPrev: (d: T) => void): DataController<T> {
  return {
    data,
    loadMore: () => hasData(data) && setPrev(data),
  }
}

export interface List<T> {
  list: T[]
  next?: string
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not initialized
// `Loading` is loading
// `Error` is error
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

export function wrapValue<T>(
  value: Value<T>,
  setValue: (d: T | null) => void,
): ValueController<T> {
  return {
    value,
    setValue,
  }
}

/** Data is loaded, or value is set to actual value */
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

/** No value yet: value or data was just initialized */
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

/** Data is loading, or value is waiting for data */
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

/** Value is selected with some or no value, or resolved with error, or data is loaded (successfully or not) */
export function isReady<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

/** Value is selected with some or no value, or data is loaded successfully */
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

/** User explicitly set no value */
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
