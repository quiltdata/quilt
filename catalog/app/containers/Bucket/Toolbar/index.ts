export * from './types'
export * from './Toolbar'
