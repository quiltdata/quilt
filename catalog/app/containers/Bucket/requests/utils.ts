import * as R from 'ramda'

export const decodeS3Key = R.pipe(R.replace(/\+/g, ' '), decodeURIComponent)
