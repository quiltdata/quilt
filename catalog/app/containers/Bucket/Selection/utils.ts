import { join } from 'path'

import * as R from 'ramda'

import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export interface SelectionItem {
  logicalKey: string
}

export interface ListingSelection {
  [prefixUrl: string]: SelectionItem[]
}

export const EMPTY_MAP: ListingSelection = {}

interface SelectionHandles {
  [prefixUrl: string]: Model.S3.S3ObjectLocation[]
}

const convertIdToHandle = (
  id: string | number,
  parentHandle: Model.S3.S3ObjectLocation,
): Model.S3.S3ObjectLocation => ({
  bucket: parentHandle.bucket,
  key: join(parentHandle.key, id.toString()),
})

export const toHandlesMap = (selection: ListingSelection): SelectionHandles =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => ({
      ...memo,
      [prefixUrl]: items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    }),
    {} as SelectionHandles,
  )

export const toHandlesList = (selection: ListingSelection): Model.S3.S3ObjectLocation[] =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => [
      ...memo,
      ...items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    ],
    [] as Model.S3.S3ObjectLocation[],
  )

const mergeWithFiltered =
  (prefix: string, filteredItems: SelectionItem[]) => (allItems: SelectionItem[]) => {
    if (!allItems || !allItems.length) return filteredItems
    const selectionOutsideFilter = allItems.filter(
      (item) => !item.logicalKey.startsWith(prefix),
    )
    const newIds = [...selectionOutsideFilter, ...filteredItems]
    return R.equals(newIds, allItems) ? allItems : newIds // avoids cyclic update
  }

export function merge(
  items: SelectionItem[],
  bucket: string,
  path: string,
  filter?: string,
): (state: ListingSelection) => ListingSelection {
  const prefixUrl = `s3://${bucket}/${path}`
  const lens = R.lensProp<Record<string, SelectionItem[]>>(prefixUrl)
  return filter ? R.over(lens, mergeWithFiltered(filter, items)) : R.set(lens, items)
}

const EmptyKeys: string[] = []

export const getDirectorySelection = (
  selection: ListingSelection,
  bucket: string,
  path: string,
) => selection[`s3://${bucket}/${path}`] || EmptyKeys
