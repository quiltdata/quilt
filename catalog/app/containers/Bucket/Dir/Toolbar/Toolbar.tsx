import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import cfg from 'constants/config'
import * as BucketPreferences from 'utils/BucketPreferences'

import * as Toolbar from 'containers/Bucket/Toolbar'
import { usePackageCreationDialog } from 'containers/Bucket/PackageDialog'
import * as Selection from 'containers/Bucket/Selection'
import ToolbarErrorBoundary from 'containers/Bucket/Toolbar/ErrorBoundary'

import * as Add from './Add'
import * as CreatePackage from './CreatePackage'
import * as Get from './Get'
import * as Organize from './Organize'

export { DirHandleCreate as CreateHandle } from 'containers/Bucket/Toolbar'
export { Add, CreatePackage, Get, Organize }

const useStyles = M.makeStyles((t) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    gap: t.spacing(1),
  },
}))

interface Features {
  add: boolean | null
  get: false | { code: boolean } | null
  organize: boolean | null
  createPackage: boolean | null
}

export function useFeatures(): Features | null {
  const { prefs } = BucketPreferences.use()
  return React.useMemo(
    () =>
      BucketPreferences.Result.match(
        {
          Ok: ({ ui: { actions, blocks } }) => ({
            add: actions.writeFile,
            get:
              !cfg.noDownload && actions.downloadObject ? { code: blocks.code } : false,
            organize: true,
            createPackage: actions.createPackage,
          }),
          _: () => null,
        },
        prefs,
      ),
    [prefs],
  )
}

interface DirToolbarProps {
  className?: string
  features: Features | null
  handle: Toolbar.DirHandle
  onReload: () => void
}

function DirToolbar({ className, features, handle, onReload }: DirToolbarProps) {
  const classes = useStyles()
  const slt = Selection.use()

  const { path, bucket } = handle

  const packageDirectoryDialog = usePackageCreationDialog({
    s3Path: path,
    bucket,
    delayHashing: true,
    disableStateDisplay: true,
  })

  const openPackageCreationDialog = React.useCallback(
    (successor) => {
      packageDirectoryDialog.open({
        path,
        selection: slt.selection,
        successor,
      })
    },
    [packageDirectoryDialog, path, slt.selection],
  )

  const successors = CreatePackage.useSuccessors(bucket)

  if (!features)
    return (
      <div className={cx(classes.root, className)}>
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
      </div>
    )

  return (
    <div className={cx(classes.root, className)}>
      <ToolbarErrorBoundary>
        {packageDirectoryDialog.render({
          successTitle: 'Package created',
          successRenderMessage: ({ packageLink }) => (
            <>Package {packageLink} successfully created</>
          ),
          title: 'Create package',
        })}

        {features.add && (
          <Add.Context.Provider handle={handle}>
            <Toolbar.Add>
              <Add.Options />
            </Toolbar.Add>
          </Add.Context.Provider>
        )}

        {features.get && (
          <Toolbar.Get>
            <Get.Options handle={handle} hideCode={!features.get.code} />
          </Toolbar.Get>
        )}

        {features.organize && (
          <Organize.Context.Provider onReload={onReload}>
            <Toolbar.Organize>
              <Organize.Options />
            </Toolbar.Organize>
          </Organize.Context.Provider>
        )}

        {features.createPackage && (
          <Toolbar.CreatePackage>
            <CreatePackage.Options
              onChange={openPackageCreationDialog}
              successors={successors}
            />
          </Toolbar.CreatePackage>
        )}
      </ToolbarErrorBoundary>
    </div>
  )
}

export { DirToolbar as Toolbar }
