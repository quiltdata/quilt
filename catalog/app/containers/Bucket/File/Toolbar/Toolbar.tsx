import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import type { EditorState } from 'components/FileEditor'
import * as Toolbar from 'containers/Bucket/Toolbar'
import type { ViewModes } from 'containers/Bucket/viewModes'
import cfg from 'constants/config'
import * as BucketPreferences from 'utils/BucketPreferences'
import ToolbarErrorBoundary from 'containers/Bucket/Toolbar/ErrorBoundary'

import * as Get from './Get'
import * as Organize from './Organize'

export { FileHandleCreate as CreateHandle } from 'containers/Bucket/Toolbar'
export { Get, Organize }

interface Features {
  get: false | { code: boolean } | null
  organize: boolean | null
  qurator: boolean | null
}

export function useFeatures(deleted?: boolean): Features | null {
  const { prefs } = BucketPreferences.use()
  return BucketPreferences.Result.match(
    {
      Ok: ({ ui: { actions, blocks } }) => ({
        get:
          !deleted && !cfg.noDownload && actions.downloadObject
            ? { code: blocks.code }
            : false,
        organize: true,
        qurator: blocks.qurator,
      }),
      _: () => null,
    },
    prefs,
  )
}

const useStyles = M.makeStyles((t) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    gap: t.spacing(1),
  },
  qurator: {
    marginLeft: t.spacing(-1),
  },
}))

interface FileToolbarProps {
  children?: React.ReactNode
  className?: string
  deleted?: boolean
  editorState?: EditorState
  features: Features | null
  handle: Toolbar.FileHandle
  onReload: () => void
  viewModes: ViewModes
}

export function FileToolbar({
  children,
  className,
  editorState,
  features,
  handle,
  onReload,
  viewModes,
}: FileToolbarProps) {
  const classes = useStyles()

  if (!features)
    return (
      <div className={cx(classes.root, className)}>
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
      </div>
    )

  return (
    <div className={cx(classes.root, className)}>
      <ToolbarErrorBoundary>
        {children}

        {features.get && (
          <Toolbar.Get label="Get file">
            <Get.Options handle={handle} hideCode={!features.get.code} />
          </Toolbar.Get>
        )}

        {features.organize && !!editorState && (
          <Organize.Context.Provider
            editorState={editorState}
            handle={handle}
            onReload={onReload}
          >
            <Toolbar.Organize>
              <Organize.Options viewModes={viewModes} />
            </Toolbar.Organize>
          </Organize.Context.Provider>
        )}

        {features.qurator && <Toolbar.Assist className={classes.qurator} />}
      </ToolbarErrorBoundary>
    </div>
  )
}

export { FileToolbar as Toolbar }
