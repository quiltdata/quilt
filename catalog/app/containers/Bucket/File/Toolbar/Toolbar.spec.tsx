import * as React from 'react'
import { render } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'

import * as BucketPreferences from 'utils/BucketPreferences'
import { extendDefaults } from 'utils/BucketPreferences/BucketPreferences'

import * as FileToolbar from './Toolbar'

jest.mock('constants/config', () => ({}))

jest.mock('./Get', () => ({
  Options: () => <div>"Get" popover</div>,
}))

jest.mock('./Organize', () => ({
  Context: {
    Provider: ({ children }: { children: React.ReactNode }) => <>{children}</>,
  },
  Options: () => <>"Organize" popover</>,
}))

jest.mock('@material-ui/lab', () => ({
  ...jest.requireActual('@material-ui/lab'),
  Skeleton: () => <i>âŒ›</i>,
}))

jest.mock('components/Buttons', () => ({
  ...jest.requireActual('components/Buttons'),
  WithPopover: ({
    label,
    children,
    disabled,
  }: {
    disabled: boolean
    label: string
    children: React.ReactNode
  }) => (
    <button title={label} disabled={disabled}>
      {children}
    </button>
  ),
}))

jest.mock('containers/Bucket/Toolbar', () => ({
  ...jest.requireActual('containers/Bucket/Toolbar'),
  Assist: () => <button>Assist</button>,
}))

const prefsHook: jest.Mock<{ prefs: BucketPreferences.Result }> = jest.fn(() => ({
  prefs: BucketPreferences.Result.Init(),
}))

jest.mock('utils/BucketPreferences', () => ({
  ...jest.requireActual('utils/BucketPreferences'),
  use: () => prefsHook(),
}))

const viewModes = { modes: [], mode: null, handlePreviewResult: jest.fn() }

const editorState = {
  editing: null,
  error: null,
  onCancel: jest.fn(),
  onChange: jest.fn(),
  onEdit: jest.fn(),
  onPreview: jest.fn(),
  onSave: jest.fn(),
  preview: false,
  saving: false,
  types: [],
  value: 'test content',
}

describe('useFeatures', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should return null when preferences are loading', () => {
    prefsHook.mockImplementationOnce(() => ({
      prefs: BucketPreferences.Result.Pending(),
    }))

    const { result } = renderHook(() => FileToolbar.useFeatures())

    expect(result.current).toBeNull()
  })

  it('should return all features disabled when all permissions are false', () => {
    prefsHook.mockImplementationOnce(() => ({
      prefs: BucketPreferences.Result.Ok(
        extendDefaults({
          ui: {
            actions: {
              downloadObject: false,
            },
            blocks: {
              code: false,
              qurator: false,
            },
          },
        }),
      ),
    }))

    const { result } = renderHook(() => FileToolbar.useFeatures(false))

    expect(result.current).toEqual({
      get: false,
      organize: true, // Always true in the implementation
      qurator: false,
    })
  })

  it('should return all features enabled when all permissions are true', () => {
    prefsHook.mockImplementationOnce(() => ({
      prefs: BucketPreferences.Result.Ok(
        extendDefaults({
          ui: {
            actions: {
              downloadObject: true,
            },
            blocks: {
              code: true,
              qurator: true,
            },
          },
        }),
      ),
    }))

    const { result } = renderHook(() => FileToolbar.useFeatures(false))

    expect(result.current).toEqual({
      get: { code: true },
      organize: true,
      qurator: true,
    })
  })

  it('should return get as false when file is deleted', () => {
    prefsHook.mockImplementationOnce(() => ({
      prefs: BucketPreferences.Result.Ok(
        extendDefaults({
          ui: {
            actions: {
              downloadObject: true,
            },
            blocks: {
              code: true,
              qurator: true,
            },
          },
        }),
      ),
    }))

    const { result } = renderHook(() => FileToolbar.useFeatures(true))

    expect(result.current).toEqual({
      get: false,
      organize: false,
      qurator: true,
    })
  })
})

const handle = FileToolbar.CreateHandle('test-bucket', 'test/file.txt')

describe('Toolbar', () => {
  it('should render skeleton buttons when features is null', () => {
    const { container } = render(
      <FileToolbar.Toolbar
        features={null}
        handle={handle}
        onReload={jest.fn()}
        viewModes={viewModes}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render all buttons when all features are enabled', () => {
    const { container } = render(
      <FileToolbar.Toolbar
        features={{ get: { code: true }, organize: true, qurator: true }}
        handle={handle}
        onReload={jest.fn()}
        viewModes={viewModes}
        editorState={editorState}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render nothing when all features are disabled', () => {
    const { container } = render(
      <FileToolbar.Toolbar
        features={{ get: false, organize: false, qurator: false }}
        handle={handle}
        onReload={jest.fn()}
        viewModes={viewModes}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render buttons for enabled features: get, qurator', () => {
    const { container } = render(
      <FileToolbar.Toolbar
        features={{ get: { code: true }, organize: false, qurator: true }}
        handle={handle}
        onReload={jest.fn()}
        viewModes={viewModes}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should not render organize button when editorState is not provided', () => {
    const { container } = render(
      <FileToolbar.Toolbar
        features={{ get: false, organize: true, qurator: false }}
        handle={handle}
        onReload={jest.fn()}
        viewModes={viewModes}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
