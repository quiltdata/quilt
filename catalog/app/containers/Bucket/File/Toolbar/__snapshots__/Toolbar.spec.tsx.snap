// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar should not render organize button when editorState is not provided 1`] = `
<div
  class="makeStyles-root-10"
/>
`;

exports[`Toolbar should render all buttons when all features are enabled 1`] = `
<div
  class="makeStyles-root-3"
>
  <button
    title="Get file"
  >
    <div>
      "Get" popover
    </div>
  </button>
  <span
    class="MuiBadge-root"
  >
    <button
      title="Organize"
    >
      "Organize" popover
    </button>
    <span
      class="MuiBadge-badge makeStyles-badge-5 MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary MuiBadge-invisible"
    >
      0
    </span>
  </span>
  <button>
    Assist
  </button>
</div>
`;

exports[`Toolbar should render buttons for enabled features: get, qurator 1`] = `
<div
  class="makeStyles-root-8"
>
  <button
    title="Get file"
  >
    <div>
      "Get" popover
    </div>
  </button>
  <button>
    Assist
  </button>
</div>
`;

exports[`Toolbar should render nothing when all features are disabled 1`] = `
<div
  class="makeStyles-root-6"
/>
`;

exports[`Toolbar should render skeleton buttons when features is null 1`] = `
<div
  class="makeStyles-root-1"
>
  <i>
    ⌛
  </i>
  <i>
    ⌛
  </i>
  <i>
    ⌛
  </i>
</div>
`;
